src/lib/Microsoft.Health.Fhir.Ingest/Telemetry/NormalizationExceptionTelemetryProcessor.cs

```diff
@@ -5,30 +5,50 @@
 
 using System;
 using System.Linq;
-using EnsureThat;
+using EnsureThat;
+using Microsoft.Health.Common.Errors;
 using Microsoft.Health.Common.Telemetry;
 using Microsoft.Health.Fhir.Ingest.Template;
 using Microsoft.Health.Logging.Telemetry;
-
+using Newtonsoft.Json.Linq;
+
 namespace Microsoft.Health.Fhir.Ingest.Telemetry
 {
     public class NormalizationExceptionTelemetryProcessor : ExceptionTelemetryProcessor
     {
         private readonly string _connectorStage = ConnectorOperation.Normalization;
         private static readonly Type[] DefaultExceptions = new[] { typeof(IncompatibleDataException) };
+        private IErrorMessageService _errorMessageService;
 
         public NormalizationExceptionTelemetryProcessor(params Type[] handledExceptionTypes)
             : base(handledExceptionTypes.Union(DefaultExceptions).ToArray())
         {
+        }
+
+        public NormalizationExceptionTelemetryProcessor(IErrorMessageService errorMessageService, params Type[] handledExceptionTypes)
+            : base(handledExceptionTypes.Union(DefaultExceptions).ToArray())
+        {
+            _errorMessageService = errorMessageService;
         }
 
-        public override bool HandleException(Exception ex, ITelemetryLogger logger)
+        public override bool HandleException(Exception ex, JToken message, ITelemetryLogger logger)
         {
             EnsureArg.IsNotNull(ex, nameof(ex));
             EnsureArg.IsNotNull(logger, nameof(logger));
 
             var exceptionTypeName = ex.GetType().Name;
-            return HandleException(ex, logger, IomtMetrics.HandledException(exceptionTypeName, _connectorStage), IomtMetrics.UnhandledException(exceptionTypeName, _connectorStage));
+            var shouldContinue = HandleException(ex, logger, IomtMetrics.HandledException(exceptionTypeName, _connectorStage), IomtMetrics.UnhandledException(exceptionTypeName, _connectorStage));
+
+            if (_errorMessageService != null)
+            {
+                var errorMessage = new ErrorMessage();
+                errorMessage.InputMessage = message;
+                errorMessage.Details = ex.Message;
+                errorMessage.Type = exceptionTypeName;
+                _errorMessageService.ReportError(errorMessage, default);
```

Comment:
Are we going to report errors here regardless of whether or not the exception is handled? I'm just wondering if that might cause confusion, especially if there is no action that is required by the customer.

Reply:
Well examples of handled exceptions would be invalid patient id, duplicate patient id, bad data in the device message, etc.  Handled just means we don't block the pipeline and transform it to telemetry.  Most non-handled exceptions are actually ones that we wouldn't necessary emit error messages for (cannot connect to the Event Hub, incorrect permissions, invalid FHIR server).  These are the ones we retry.

Reply:
Yeah I was on the fence about this. We should probably only report errors that are handled like patient id not found, duplicate patient id, etc, so I can update the code to only emit error messages for "handled" exceptions. 

I do think there is some value for alerting the customer on non-handled errors like (cannot connect to Event Hub, invalid FHIR server, etc if they are not transient) and I do think there is some value for having one location that contains all the errors, but it can be solved another way like logs or portal warnings.

For this project it is probably best to have actionable dead letter errors only.

Reply:
It makes me wonder if error reporting is enabled, the term 'handled' is obsolete. If an exception is encountered, do we always just report the error and move on?

Reply:
I'm still confused about what we are doing here. I thought one of the main reasons we are adding the Error Message Service is to eliminate the possibility of the pipeline being blocked by a poison message. The current implementation will still allow that to happen, but flood the Error Message Service with the same error every time it happens.

In my mind there is no such thing as a "handled" exception anymore when Error Message Service is enabled; either a message is successfully processed or it isn't. If a message isn't successfully processed, it should result in an error message. The problem with labeling an error as "transient" is we don't know that an issue was transient until we look back. If an error could be transient, I would propose that we retry a fixed number of times but not indefinitely. **The indefinite retry is where we run into the possibility of the pipeline being blocked and in an unrecoverable state, which will result in data loss.**

Reply:
I think the first step is exposing the error to users if it is actionable to them, and in parallel we should be updating the code to throw appropriate exceptions so that we know when to retry and when not to retry, and when to expose the error to the error message service. We currently assume that if the exception is "unknown" or "unhandled" then we should retry forever. One partial solution could be to update our code - for example we recently added a "NormalizationDataMappingException" which we can now use to determine if we can immediately log an error and move on. Previously we retried forever and this was a "poisoned message" scenario but now it is not.

Another solution for "unknown" or "unhandled" exceptions is, liked you mentioned, to retry X number of times and then log an error. That can work in a number of cases especially if the error is with the message itself. One case where it may be ambiguous is with connectivity errors like HTTP 500 or a managed identity related issue. Blips happen and I think we should retry forever in this scenario and attempt to fix the underlying connectivity issue, rather than retrying X number of times and logging an error, since the user can't really fix the error. On the other hand, if the connectivity issue is not resolved in a timely manner and the data fades out then it could result in data loss.

imo we should update the various places in our code to be able to detect a poisoned message and throw an exception like in the case of the NormalizationDataMappingException. And in the case where the exception is still not known then we can retry X times and log an error, but need to take into consideration connectivity related errors and handle those appropriately

Reply:
I think it's fine retrying connectivity issues (connectivity to our internal resources only) for a specific time period (maybe 15 minutes), and then sending an error to the Error Message Service so the potential of data loss is reduced, but are there other scenarios that would warrant retrying? I can't think of anything that isn't under the customer's control.